xsnap/sources/xsnap-worker.c

```diff
@@ -93,14 +93,28 @@ xsCallback gxSnapshotCallbacks[mxSnapshotCallbackCount] = {
 	// fx_clearTimer,
 };
 
+typedef struct {
+	FILE *file;
+	int size;
+} SnapshotStream;
+
 static int fxSnapshotRead(void* stream, void* address, size_t size)
 {
-	return (fread(address, size, 1, stream) == 1) ? 0 : errno;
+	SnapshotStream* snapshotStream = stream;
+	if (snapshotStream->size >= 0 && (signed)(snapshotStream->size - size) < 0)
```

Comment:
FWIW, I believe the snapshot format is self-delimiting: unless something is confused or corrupted, the `fxSnapshotRead()` call should exactly consume the last byte of the snapshot, and it will never read beyond the end.

Reply:
BTW the `(signed)` cast is a little bit unnerving. I think the logic is correct, but it might be more obvious if you used a separate boolean flag to mean "we know a size", use `size_t remaining` instead of `int size`, and then do purely unsigned comparisons like `if (snapshotStream->hasSize && size > snapshotStream->remaining) return EIO;`

Reply:
Yeah since we're not using this "EOF" style check, I might rip it out entirely.